flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java

```diff
@@ -22,65 +22,89 @@
 import java.util.List;
 import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
 
 /**
  * Base source operator only used for integrating the source reader which is proposed by FLIP-27. It implements
  * the interface of {@link PushingAsyncDataInput} for naturally compatible with one input processing in runtime
  * stack.
  *
- * <p>Note: We are expecting this to be changed to the concrete class once SourceReader interface is introduced.
+ * <p><b>Important Note on Serialization:</b> The SourceOperator inherits the {@link java.io.Serializable}
+ * interface from the StreamOperator, but is in fact NOT serializable. The operator must only be instantiates
+ * in the StreamTask from its factory.
  *
  * @param <OUT> The output type of the operator.
  */
 @Internal
+@SuppressWarnings("serial")
 public class SourceOperator<OUT, SplitT extends SourceSplit>
 		extends AbstractStreamOperator<OUT>
 		implements OperatorEventHandler, PushingAsyncDataInput<OUT> {
+
 	// Package private for unit test.
 	static final ListStateDescriptor<byte[]> SPLITS_STATE_DESC =
 			new ListStateDescriptor<>("SourceReaderState", BytePrimitiveArraySerializer.INSTANCE);
 
-	private final Source<OUT, SplitT, ?> source;
+	/** The factory for the source reader. This is a workaround, because currently the SourceReader
+	 * must be lazily initialized, which is mainly because the metrics groups that the reader relies on is
+	 * lazily initialized. */
+	private final Function<SourceReaderContext, SourceReader<OUT, SplitT>> readerFactory;
+
+	/** The serializer for the splits, applied to the split types before storing them in the reader state. */
+	private final SimpleVersionedSerializer<SplitT> splitSerializer;
+
+	/** The event gateway through which this operator talks to its coordinator. */
+	private final OperatorEventGateway operatorEventGateway;
+
+	// ---- lazily initialized fields ----
+
+	/** The source reader that does most of the work. */
+	private SourceReader<OUT, SplitT> sourceReader;
+
+	/** The state that holds the currently assigned splits. */
+	private ListState<SplitT> readerState;
 
-	// Fields that will be setup at runtime.
-	private transient SourceReader<OUT, SplitT> sourceReader;
-	private transient SimpleVersionedSerializer<SplitT> splitSerializer;
-	private transient ListState<byte[]> readerState;
-	private transient OperatorEventGateway operatorEventGateway;
+	public SourceOperator(
+			OperatorID operatorId,
+			Function<SourceReaderContext, SourceReader<OUT, SplitT>> readerFactory,
+			OperatorEventDispatcher eventDispatcher,
+			SimpleVersionedSerializer<SplitT> splitSerializer) {
 
-	public SourceOperator(Source<OUT, SplitT, ?> source) {
-		this.source = source;
+		this.readerFactory = checkNotNull(readerFactory);
+		this.splitSerializer = checkNotNull(splitSerializer);
+
+		this.operatorEventGateway = eventDispatcher.registerEventHandler(operatorId, this);
```

Comment:
Passing `this` out of a constructor is in general dangerous. The following article gives a very similar example.
https://www.ibm.com/developerworks/java/library/j-jtp0618/index.html#2

Reply:
Generally, I agree with that. I think having it as a "last step" in the constructor can be fine, because the object is constructed then. Downside is that it makes the class unsafe for inheritance (outside tests), which may be fine here.

In this case, we could also register the listener in the `open()` method - but that has the drawback of having a partially non-initialized operator, which is also not great.